### PR TITLE
tests: only use -proposed for ua (SC-270)

### PR DESCRIPTION
**Pull request by orndorffgrant**
## Proposed Commit Message

```
tests: only use -proposed for ua
```

## Test Steps

This is hard to test now that our latest release is in -updates. I suggest the following in a fresh xenial container:
1. Create the files that the new `printf` commands in this PR are creating
2. Change the specific package in `/etc/apt/preferences.d/uaclient-proposed` to `autopkgtest`
3. `apt update`
4. `apt install autopkgtest`
5. `apt policy autopkgtest`
    * Ensure that -proposed is pinned at 400 but this package is pinned at 1001. Make sure the version of autopkgtest from proposed is installed
6. `apt upgrade`
    * Ensure that nothing else is installed from -proposed

Also, still run `env UACLIENT_BEHAVE_ENABLE_PROPOSED=1 tox -e behave-vm-16.04 -- features/staging_commands.feature:479 --no-junit` and ensure it passes.

## Desired commit type::

 - [ ] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [x] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist:

 - [ ] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

### CODE REVIEW

File: features/environment.py [hunk 1]
```diff
@@ -848,6 +848,11 @@ def _install_uat_in_container(
             )
         )
 
+    if config.enable_proposed:
+        cmds.append(
+            ["sudo", "rm", "/etc/apt/sources.list.d/uaclient-proposed.list"]
```
Comment (renanrodrigo) [inline features/environment.py:853]: Should we run also `apt update` after this, just to make sure?

Comment (orndorffgrant) [reply features/environment.py:853]: Ah good catch - yeah we probably should.

Comment (orndorffgrant) [reply features/environment.py:853]: Just added this

**Review by renanrodrigo — APPROVED** [review]
lgtm

**Review by lucasmoura — COMMENTED** [review]
The only problem with this approach is that we will not upgrade ua during `do-release-upgrade`.
Therefore, the upgrade test will not fully exercise the scenario they were intended to test.
I think it is better for us to keep the proposed pocket enabled and if we find a problem in the upgrade tests regarding proposed, we use the ppas to avoid that issue

**Comment by orndorffgrant** [comment]
@lucasmoura Excellent point. For example, for the xenial upgrade test, xenial-proposed-ua would not get upgraded to bionic-proposed-ua. Instead, the test ends up keeping xenial-proposed-ua on bionic after the `do-release-upgrade`.

Is there any way to only enable -proposed for a single package?

**Comment by renanrodrigo** [comment]
@orndorffgrant maybe we could set cache-policies for proposal to be minimal priority, and raise it only for `ubuntu-advantage-tools`

**Comment by paride** [comment]
It would make sense if `RELEASE-proposed` somehow had lower priority then `RELEASE` when added to `sources.list`, as adding `-proposed` is mostly done to test individual packages or sets of packages. Upgrading everything from `-proposed` is in general not a good idea. Unfortunately we're not there, see:
 - https://bugs.launchpad.net/launchpad/+bug/1016776
 - https://lists.ubuntu.com/archives/ubuntu-devel/2021-January/041324.html (not implemented afaik)

The way to avoid full upgrades from `-proposed` is to pin it to a lower priority than the release: https://wiki.ubuntu.com/Testing/EnableProposed. Then packages from `-proposed` can be installed via `apt -t <target_release> install ...`

**Comment by orndorffgrant** [comment]
@lucasmoura @renanrodrigo @paride Thanks for all of the guidance! This is updated now to use the pinning strategy to enable proposed but only use it for uaclient by default. Let me know what you think

File: features/environment.py [hunk 2]
```diff
@@ -45,6 +45,8 @@
 USERDATA_RUNCMD_ENABLE_PROPOSED = """
 runcmd:
   - printf \"deb http://archive.ubuntu.com/ubuntu/ {series}-proposed restricted main multiverse universe\" > /etc/apt/sources.list.d/uaclient-proposed.list
```
Comment (paride) [inline features/environment.py:47]: `main` should be enough here, for the reset this LGTM!

**Review by paride — CHANGES_REQUESTED** [review]
One little inline comment, not on the lines added by this PR which LGTM, but let's not waste the occasion :-)

**Comment by orndorffgrant** [comment]
@paride totally agree! Just pushed a fix

**Review by paride — APPROVED** [review]
LGTM

**Review by renanrodrigo — APPROVED** [review]

**Review by lucasmoura — APPROVED** [review]

**Pull request merged**